spec/src/main/asciidoc/repository.asciidoc

```diff
@@ -112,6 +112,246 @@ Subsequently, an implementation of Jakarta Data, specifically tailored to the ch
+
+[NOTE]
+Additional portability guarantees may be provided by specifications which extend this specification, specializing to a given class of datastore.
+
+==== Lifecycle methods
+
+A _lifecycle method_ is an abstract method annotated with a _lifecycle annotation_.
+Lifecycle methods allow the program to make changes to persistent data in the data store.
+
+A lifecycle method must:
+
+- have a single parameter annotated with a lifecycle annotation, whose type is a concrete entity class `E`, and
+- be declared `void`, or return exactly the same entity type `E`.
+
+That is, its signature must be:
+
+[source,java]
+----
+void lifecycle(@Lifecycle Entity e);
+----
+
+or:
+
+[source,java]
+----
+Entity lifecycle(@Lifecycle Entity e);
+----
+
+where `lifecycle` is the arbitrary name of the method, `Entity` is a concrete entity class, and `Lifecycle` is a lifecycle annotation.
+
+This specification defines three built-in lifecycle annotations: `@Insert`, `@Update`, and `@Delete`.
+
+For example:
+
+[source,java]
+----
+void insertBook(@Insert Book book);
+----
+
+Lifecycle methods are not guaranteed to be portable between all providers.
+
+[NOTE]
+Any Jakarta Data implementation is strongly encouraged to support lifecycle methods declared using the built-in lifecycle annotations, to enable portability of user code between Jakarta Data providers.
+However, since these lifecycle operations are not universal across all possible data stores, support for `@Insert`, `@Update`, and `@Delete` is not required by this specification.
+
+There is no special programming model for lifecycle annotations.
+The Jakarta Data implementation automatically recognizes the lifecycle annotations it supports.
+
+[NOTE]
+====
+A Jakarta Data provider might extend this specification to define additional lifecycle annotations, or to support lifecycle methods with signatures other than the usual signatures defined above. For example, a provider might support "merge" methods declared as follows:
+
+[source,java]
+----
+Book mergeBook(@Merge Book book);
+----
+
+Such lifecycle methods are not portable between Jakarta Data providers.
+====
+
+==== Annotated query methods
+
+An _annotated query method_ is an abstract method annotated by a _query annotation_ type.
+The query annotation specifies a query in some datastore-native query language.
+
+Each parameter of an annotated query method must either:
+
+- have exactly the same name and type as a named parameter of the query,
+- have exactly the same type and position within the parameter list of the method as a positional parameter of the query, or
+- be of type `Limit`, `Pageable`, or `Sort`.
+
+A repository with annotated query methods with named parameters should be compiled so that parameter names are preserved in the class file (for example, using `javac -parameters`), or the parameter names must be specified explicitly using the `@Param` annotation.
```

Comment:
```suggestion
A repository with annotated query methods with named parameters must be compiled so that parameter names are preserved in the class file (for example, using `javac -parameters`), or the parameter names must be specified explicitly using the `@Param` annotation.
```